### PR TITLE
Add resource type and editor plugin for text files

### PR DESCRIPTION
**Pull request by Geequlim**
Add `RawText` to handle text files as resources.
Add RawTextEditorPlugin to edit text files with code highlight.

Supported code highlight:
- Bash Shell
- C/C++
- C#
- CSS
- JSON
- Javascript/ECMAScript 6
- Python
- INI/TOML configuration files
- XML _When `XML_ENABLED` is not defined_
- YAML

### CODE REVIEW

File: tools/editor/plugins/raw_text_editor_plugin.cpp [hunk 1]
```diff
@@ -0,0 +1,737 @@
...
+		lang_option->select(HighlightConfig::LANG_TOML);
+	else if (extension == "c++" || extension == "cxx" || extension == "cpp" || extension == "h" || extension == "hpp")
+		lang_option->select(HighlightConfig::LANG_CPP);
+	else if (extension == "py")
+		lang_option->select(HighlightConfig::LANG_PYTHON);
+	else if(extension == "js")
+		lang_option->select(HighlightConfig::LANG_JAVASCRIPT);
+	else if(extension == "cs")
+		lang_option->select(HighlightConfig::LANG_CSHARP);
+	else if(extension == "ccs")
```
Comment (Calinou) [inline tools/editor/plugins/raw_text_editor_plugin.cpp]: Shouldn't it be `else if(extension == "css")`?

Comment (Geequlim) [inline tools/editor/plugins/raw_text_editor_plugin.cpp]: Yes it should be, thanks :)

**Comment by Geequlim** [comment]
@djrm Could you please add an icon to RawText after this is merged?

File: tools/editor/plugins/raw_text_editor_plugin.cpp (hunk 1 above)
Comment (Geequlim) [inline tools/editor/plugins/raw_text_editor_plugin.cpp]: Fixed :P

**Comment by vnen** [comment]
Seems to have missed the discussion in #1390.

While this is very cool I'd rather have it in the same place of the script editor, listing the open text files along with scripts and help.

**Comment by neikeq** [comment]
The problem is that ScriptEditor is designed to work with ScriptTextEditor, which only works with scripts. However, this should be doable, but with some refactoring effort :P

**Comment by Geequlim** [comment]
@vnen Sounds great about that.

More features may be implemented in another day or by other guy because I have no enough time and energy in the few days.

**Comment by vnen** [comment]
I actually looked into this and the reason I didn't make the feature is because it's a great deal of work, but it is totally possible to refactor the ScriptEditorPlugin to allow another text editor. It already works with the help pages. I didn't do it because we're quite late for 2.1 and I wanted to focus on other things so this release can be finished.

I'm not sure if I want this PR's editor as it is, because it is kinda off-place. And there's the problem discussed in #1390 about whether it should handle any text file or just a bunch of recognized extensions. Maybe the user has their own text format for their game data and want to use the in-engine text editor.

The code highlighting from here is quite good though, and it can certainly be used if/when the text editor in script tab is implemented.

**Comment by Geequlim** [comment]
@vnen OK, I will keep this PR open before the new ScriptEditor release.

Please feel free to take code from this PR if it is usable.
2016年6月20日 23:13，"George Marques" <notifications@github.com>写道：

**Comment by Geequlim** [comment]
@vnen OK, I will keep this PR open before the new ScriptEditor release.

Please feel free to take code from this PR if it is usable.

2016年6月20日 23:36，写道：
>
> @vnen OK, I will keep this PR open before the new ScriptEditor release.
>
> Please feel free to take code from this PR if it is usable.
>
> 2016年6月20日 23:13，"George Marques" <notifications@github.com>写道：
>>
>> I actually looked into this and the reason I didn't make the feature is
because it's a great deal of work, but it is totally possible to refactor
the ScriptEditorPlugin to allow another text editor. It already works with
the help pages. I didn't do it because we're quite late for 2.1 and I
wanted to focus on other things so this release can be finished.
>>
>> I'm not sure if I want this PR's editor as it is, because it is kinda
off-place. And there's the problem discussed in #1390 about whether it
should handle any text file or just a bunch of recognized extensions. Maybe
the user has their own text format for their game data and want to use the
in-engine text editor.
>>
>> The code highlighting from here is quite good though, and it can
certainly be used if/when the text editor in script tab is implemented.

**Comment by reduz** [comment]
Yeah, the script editor needs a refactor, and this is not the only thing that has to change.. should do a document specifying everything that has to change

File: tools/editor/plugins/raw_text_editor_plugin.cpp [hunk 2]
```diff
@@ -0,0 +1,774 @@
...
+		}
+
+	}
+	else if( extension == "js") { // Javascript
+
+		hl.filled = true;
+
+		hl.string_delimiters.push_back("\" \"");
+		hl.string_delimiters.push_back("' '");
+		hl.string_delimiters.push_back("$` `");
```
Comment (bojidar-bg) [inline tools/editor/plugins/raw_text_editor_plugin.cpp]: Why the dollar sign <kbd>$</kbd>?

File: scene/resources/raw_text.cpp [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+
+}
+
+void RawText::get_recognized_extensions(List<String> *p_extensions) {
+	String extensions[] = {
+		"txt", "md", "json", "toml", "cfg", "ini", "inf", "c", "cpp", "cxx",
+		"c++", "h", "hpp",  "m", "mm", "css", "htm", "html", "py", "java", "js",
+		"sh", "cmd", "csv", "cs", "as", "bash", "hlsl", "glsl", "diff", "php",
+		"lua", "rb", "yml", "yaml", "go"
+#ifndef XML_ENABLED
```
Comment (bojidar-bg) [inline scene/resources/raw_text.cpp]: You mean `XML_DISABLED`?

**Comment by akien-mga** [comment]
Discussed this one again with @reduz today, it seems it won't be merged in its current state, as it would be better to instance expand ScriptEditor to support other formats instead. Keeping it around as reference implementation and reminder that there is interest in such a feature though :)

File: scene/resources/raw_text.cpp [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	ObjectTypeDB::bind_method(_MD("get_text"),&RawText::get_text);
+	ADD_PROPERTY( PropertyInfo(Variant::STRING,"text"), _SCS("set_text"), _SCS("get_text"));
+
+}
+
+void RawText::get_recognized_extensions(List<String> *p_extensions) {
+	String extensions[] = {
+		"txt", "md", "json", "toml", "cfg", "ini", "inf", "c", "cpp", "cxx",
+		"c++", "h", "hpp",  "m", "mm", "css", "htm", "html", "py", "java", "js",
+		"sh", "cmd", "csv", "cs", "as", "bash", "hlsl", "glsl", "diff", "php",
```
Comment (neikeq) [inline scene/resources/raw_text.cpp]: This would override GodotSharp's resource loaded. Or would it be overridden by GodotSharp's?

**Comment by reduz** [comment]
Closing this, we should probably make this properly, by adding support in ScriptEditorPlugin

**Pull request closed without merging**